ztoc/toc_builder.go

```diff
@@ -0,0 +1,203 @@
+
+// TarProviderGzip creates a tar reader from gzip reader.
+func TarProviderGzip(compressedReader *os.File) (io.Reader, error) {
+	return gzip.NewReader(compressedReader)
+}
+
+// TarProviderZstd creates a tar reader from zstd reader.
+func TarProviderZstd(compressedReader *os.File) (io.Reader, error) {
+	return zstd.NewReader(compressedReader)
+}
+
+// TocBuilder builds the `TOC` part of a ztoc and works with different
+// compression algorithms (e.g., gzip, zstd) with a registered `TarProvider`.
+type TocBuilder struct {
+	tarProviders map[string]TarProvider
+}
+
+// NewTocBuilder return a `TocBuilder` struct. Users need to call `RegisterTarProvider`
+// to support a specific compression algorithm.
+func NewTocBuilder() TocBuilder {
+	return TocBuilder{
+		tarProviders: make(map[string]TarProvider),
+	}
+}
+
+// RegisterTarProvider adds a TarProvider for a compression algorithm.
+func (tb TocBuilder) RegisterTarProvider(algorithm string, provider TarProvider) {
+	if tb.tarProviders == nil {
+		tb.tarProviders = make(map[string]TarProvider)
+	}
+	tb.tarProviders[algorithm] = provider
+}
+
+// CheckCompressionAlgorithm checks if a compression algorithm is supported.
+func (tb TocBuilder) CheckCompressionAlgorithm(algorithm string) bool {
+	_, ok := tb.tarProviders[algorithm]
+	return ok
+}
+
+// TocFromFile creates a `TOC` given a layer blob filename and the compression
+// algorithm used by the layer.
+func (tb TocBuilder) TocFromFile(algorithm, filename string) (TOC, compression.Offset, error) {
+	if !tb.CheckCompressionAlgorithm(algorithm) {
+		return TOC{}, 0, fmt.Errorf("unsupported compression algorithm: %s", algorithm)
+	}
+
+	fm, uncompressedArchiveSize, err := tb.getFileMetadata(algorithm, filename)
+	if err != nil {
+		return TOC{}, 0, err
+	}
+
+	return TOC{Metadata: fm}, uncompressedArchiveSize, nil
+}
+
+// getFileMetadata creates `FileMetadata` for each file within the compressed file
+// and calculate the uncompressed size of the passed file.
+func (tb TocBuilder) getFileMetadata(algorithm, filename string) ([]FileMetadata, compression.Offset, error) {
+	// read compress file and create compress tar reader.
+	compressFile, err := os.Open(filename)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not open file for reading: %v", err)
+	}
+	defer compressFile.Close()
+
+	compressTarReader, err := tb.tarProviders[algorithm](compressFile)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// create uncompress temp tar file and uncompress tar reader.
+	uncompressFile, err := os.CreateTemp("/tmp", "tempfile-ztoc-builder")
+	if err != nil {
+		return nil, 0, err
+	}
+	defer os.Remove(uncompressFile.Name())
+
+	uncompressFileSize, err := io.Copy(uncompressFile, compressTarReader)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// create toc from tar reader.
+	tarSectionReader := io.NewSectionReader(uncompressFile, 0, uncompressFileSize)
+	md, err := metadataFromTarReader(tarSectionReader)
```

Comment:
I didn't change this `create .tar temp file then read from it as the tar reader` to keep it consistent with existing impl. But I think we can just pass the `compressTarReader` to `metadataFromTarReader` (as the arg for `tar.NewReader`), and let `metadataFromTarReader` returns the `uncompressFileSize` as well since it has `positionTrackerReader`.

This can  avoid creating yet another temp `.tar`. (in soci index create we create a temp `.tar.gz` as input to ztoc builder)